### PR TITLE
Support multiple optimMethods in TFTrainingHelper

### PR DESCRIPTION
**Pull request by hkvision**
- Train a TF unfreeze graph using multiple optim methods based on scope
- Bug fix for Python Estimator.
- Unit tests.

### CODE REVIEW

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/net/TFTrainingHelper.scala [hunk 1]
```diff
@@ -57,6 +57,27 @@ private[zoo] class TFTrainingHelper(tfnet: TFNet,
...
+    val map = collection.mutable.Map[String, Tensor[Float]]()
+    var i = 0
+    while (i < variables.length) {
+      map(variables(i)) = weights(i)
+      i +=1
+    }
+    map
+  }
+
+  override def apply(name: String): Option[AbstractModule[Activity, Activity, Float]] = {
```
Comment (yangw1234) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/net/TFTrainingHelper.scala]: Should we make sure that each `TFSubGraph`'s weights' underlying storage is a continuous subset of the `TFTrainingHelper`'s weights' storage.

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/net/TFTrainingHelper.scala [hunk 2]
```diff
@@ -57,6 +57,27 @@ private[zoo] class TFTrainingHelper(tfnet: TFNet,
...
+    val targetVariables = if (name == getName()) variables else variables.filter(_.contains(name))
```
Comment (yangw1234) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/net/TFTrainingHelper.scala]: Should we right `val targetVariables = variables.filter(_.startsWith(name))`.

Say we have variables namely "a/b/c", "c/a/b" and "a". The sub-graph "a" should contains "a" and "a/b/c".

Comment (hkvision) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/net/TFTrainingHelper.scala]: Sure.

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/net/TFTrainingHelper.scala (hunk 1 above)
Comment (hkvision) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/net/TFTrainingHelper.scala]: https://github.com/intel-analytics/BigDL/blob/master/spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Optimizer.scala#L504
It is checked in Optimizer? But anyway, it will be safer to check it here.

Comment (hkvision) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/net/TFTrainingHelper.scala]: @yangw1234 Seems that the most secure way to do the check is via `checkSubModules` in Optimizer, if I write by myself, does the following covers all cases?
```
val targetStorage = targetWeights.flatMap(x => x.storage().array().slice(x.storageOffset(), x.storageOffset() + x.nElement()))
val totalStorage = Module.flatten(parameters()._1).storage().array()
require(totalStorage.containsSlice(targetStorage))
```
Take a look, thanks.

**Comment by hkvision** [comment]
Can move the ut to Python if necessary. Python API doesn't change.

**Comment by hkvision** [comment]
jenkins passed http://10.239.47.211:18888/job/ZOO-PR-Validation/2530/

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/net/TFTrainingHelper.scala [hunk 3]
```diff
@@ -144,6 +162,21 @@ private[zoo] class TFTrainingHelper(tfnet: TFNet,
...
+  override def updateOutput(input: Activity): Activity = {
+    input
+  }
+
+  override def updateGradInput(input: Activity, gradOutput: Activity): Activity = {
+    gradInput
+  }
+
+  override def parameters(): (Array[Tensor[Float]], Array[Tensor[Float]]) = {
+    (weights, weights.map(_ => Tensor[Float]()))
```
Comment (qiuxin2012) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/net/TFTrainingHelper.scala]: grad's length should be the same with weights

**Comment by hkvision** [comment]
Need to sort variable names otherwise weights under the same scope are not contiguous.

**Comment by hkvision** [comment]
Sorted variables by name @yangw1234

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/net/TFTrainingHelper.scala [hunk 4]
```diff
@@ -58,6 +56,27 @@ private[zoo] class TFTrainingHelper(tfnet: TFNet,
     setWeights(ws)
   }
 
+  private val weightsMap = {
```
Comment (yangw1234) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/net/TFTrainingHelper.scala]: You should change TFTrainingHelper2 which is the not-freeze implementation

File: pyzoo/zoo/pipeline/api/net/tf_optimizer.py [hunk 5]
```diff
@@ -562,6 +563,7 @@ def from_keras(cls, keras_model, dataset, optim_method=None, val_spilt=0.0, **kw
         inputs = model_inputs + model_targets
 
         variables = keras_model._collected_trainable_weights
+        variables.sort(key=lambda variable: variable.name)
```
Comment (yangw1234) [inline pyzoo/zoo/pipeline/api/net/tf_optimizer.py]: do this in the constructor? If we add other factory methods to create TFOptimizer, we might forget to sort the variables.

File: zoo/src/test/scala/com/intel/analytics/zoo/pipeline/api/net/TFNetSpec.scala [hunk 6]
```diff
@@ -160,4 +164,36 @@ class TFNetSpec extends FlatSpec with Matchers with BeforeAndAfter {
...
+      i += 1
+    }
+
+    val rdd = sc.parallelize(data)
+    val dataSet = DataSet.rdd(rdd) -> SampleToMiniBatch[Float](128)
+    val optimizer = new DistriOptimizer[Float](
+      layer,
+      dataSet.asInstanceOf[DistributedDataSet[MiniBatch[Float]]],
+      new IdentityCriterion())
+      .setOptimMethods(
```
Comment (yangw1234) [inline zoo/src/test/scala/com/intel/analytics/zoo/pipeline/api/net/TFNetSpec.scala]: I think we should enable user to set it TFOptimizer, KerasModel and TFEstimator.

Comment (hkvision) [reply zoo/src/test/scala/com/intel/analytics/zoo/pipeline/api/net/TFNetSpec.scala]: Supported in TFOptimizer and TFEstimator. Pending for KerasModel.

File: pyzoo/zoo/pipeline/api/net/tf_optimizer.py (hunk 5 above)
Comment (hkvision) [reply pyzoo/zoo/pipeline/api/net/tf_optimizer.py]: Added remarks. There is little possibility that we have other factory methods.

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/net/TFTrainingHelper.scala (hunk 4 above)
Comment (hkvision) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/net/TFTrainingHelper.scala]: Changed.

File: pyzoo/test/zoo/tfpark/test_tfpark_tfoptimizer.py [hunk 7]
```diff
@@ -66,15 +65,22 @@ def test_tf_optimizer_metrics(self):
...
                                               val_outputs=[output],
                                               val_labels=[label_tensor],
                                               val_method=Accuracy(), metrics={"loss": loss})
+            initial_weights = optimizer.tf_model.training_helper_layer.get_weights()
             optimizer.optimize(end_trigger=MaxEpoch(1))
+            updated_weights = optimizer.tf_model.training_helper_layer.get_weights()
+            for i in [0, 1]:  # weights and bias combined with "dense/" should be updated
+                assert not np.allclose(initial_weights[i], updated_weights[i])
+            for i in [2, 3]:  # weights and bias combined with "dense_1" should be unchanged
+                assert np.allclose(initial_weights[i], updated_weights[i])
```
Comment (hkvision) [inline pyzoo/test/zoo/tfpark/test_tfpark_tfoptimizer.py]: Added verification of weights @yangw1234

**Comment by hkvision** [comment]
Please take a look if there are further comments. @yangw1234

**Comment by hkvision** [comment]
jenkins passed: http://10.239.47.211:18888/job/ZOO-PR-Validation/2853/

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/net/TFTrainingHelper2.scala [hunk 8]
```diff
@@ -304,3 +327,17 @@ case class TrainMeta2(inputs: Array[String],
...
+  override def updateOutput(input: Activity): Activity = {
+    input
+  }
+
+  override def updateGradInput(input: Activity, gradOutput: Activity): Activity = {
+    gradInput
+  }
+
+  override def parameters(): (Array[Tensor[Float]], Array[Tensor[Float]]) = {
+    (weights, weights.map(_ => Tensor[Float]()))
```
Comment (yangw1234) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/net/TFTrainingHelper2.scala]: Nitpick: One might wonder why the gradients of `TFSubGraph` does not have to be the same instances of the original `TFTrainingHelper`'s. The truth is they are never used in DistriOptimizer. And the DistriOptimizer only use the weights' size information to slice on the original weights and gradients. It is better leave some comments about that. Also, this way, if the implementation of DistriOptimizer changed, may lead to incorrect result.

**Comment by yangw1234** [comment]
LGTM

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/net/TFTrainingHelper2.scala (hunk 8 above)
Comment (hkvision) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/net/TFTrainingHelper2.scala]: Added detailed comments. If the implementation of DistriOptimizer is changed, then I suppose many many code segments would be affected :)

**Comment by hkvision** [comment]
Thanks for your review @yangw1234

**Comment by hkvision** [comment]
jenkins passed http://10.239.47.211:18888/job/ZOO-PR-Validation/2855/
